{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 1 File connection and reading the first lines\n",
    "\n",
    "Open a connection to the file: Bolivian_squirrel_monkey_genes_mart_export.txt to read the data in the file. \n",
    "\n",
    "To learn what data is stored and how: Read and print the first ten lines from the file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "bsm_file = open(\"Bolivian_squirrel_monkey_genes_mart_export.txt\",\"r\")\n",
    "\n",
    "print(bsm_file.readline())\n",
    "print(bsm_file.readline())\n",
    "print(bsm_file.readline())\n",
    "print(bsm_file.readline())\n",
    "print(bsm_file.readline())\n",
    "print(bsm_file.readline())\n",
    "print(bsm_file.readline())\n",
    "print(bsm_file.readline())\n",
    "print(bsm_file.readline())\n",
    "print(bsm_file.readline())\n",
    "\n",
    "bsm_file.close()\n",
    "\n",
    "print(\"---------------------------------------\")\n",
    "# or (less copy paste)\n",
    "\n",
    "bsm_file = open(\"Bolivian_squirrel_monkey_genes_mart_export.txt\",\"r\")\n",
    "\n",
    "for  i in range(0,11):\n",
    "    print(bsm_file.readline())\n",
    "\n",
    "bsm_file.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 2 Getting data from a file in a list\n",
    "\n",
    "Now we know that we have lines with data that represent the header starting with >\n",
    "and the sequence (which can also be unavailable).\n",
    "\n",
    "The header contains the gene ID the start and stop position and the chromosome at which the gene can be found.\n",
    "\n",
    "The sequence can span multiple lines and when it does we need to stitch it together to obtain the complete sequence."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "2.1) First try to get the data from the file in a list an dprint the first ten items in the list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## open a connection to the file\n",
    "bsm_file = open(\"Bolivian_squirrel_monkey_genes_mart_export.txt\",\"r\")\n",
    "\n",
    "## make an empty list to store the data\n",
    "bsm_genes = []\n",
    "\n",
    "## use a loop to read all the data, line-by-line and add it to the empty list\n",
    "for line in bsm_file:\n",
    "    bsm_genes.append(line)  # you can also use: bsm_genes += [line]\n",
    "\n",
    "bsm_file.close()\n",
    "\n",
    "# print the first 10 items in the list\n",
    "print(bsm_genes[0:10])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now we have all data from the file available for Python to work with. However when we look at the frist ten items/lines we can notice some issues that are not convient.\n",
    "- the line end special character \\n is still at each line\n",
    "- the header info is concatenated, we'd like the id, start, stop and chromosome separate\n",
    "- the sequence is spread over multiple ietems/lines\n",
    "\n",
    "Let's solve this one-by-one including checks if it worked\n",
    "\n",
    "2.2) remove the end-of-line characters from the data going into the list, print the first ten items\n",
    "\n",
    "2.3) split the header info and make separate list for id, start, stop and chromosome, print the first ten items and check the length of each list\n",
    "\n",
    "2.4) concatenate the sequence-parts in to full sequences, print the first ten items and check the length of each list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 2.2) remove end-of-line characters\n",
    "\n",
    "bsm_file = open(\"Bolivian_squirrel_monkey_genes_mart_export.txt\",\"r\")\n",
    "\n",
    "bsm_genes = []\n",
    "for line in bsm_file:\n",
    "    bsm_genes.append(line.rstrip()) ## .rstrip() can be used to remove the \\n character from the end of the line\n",
    "    ## you can also use bsm_genes += [line.rstrip()]\n",
    "bsm_file.close()\n",
    "\n",
    "print(bsm_genes[0:10])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 2.3) split header and make separate list for id, start, stop and chromosome\n",
    "\n",
    "bsm_file = open(\"Bolivian_squirrel_monkey_genes_mart_export.txt\",\"r\")\n",
    "\n",
    "# we need to make a separate list for id, start, stop and chromosome and therefor also for the sequence\n",
    "bsm_id = []\n",
    "bsm_start = []\n",
    "bsm_stop = []\n",
    "bsm_chr = []\n",
    "bsm_seq = []\n",
    "\n",
    "for line in bsm_file:\n",
    "    # first remove the \\n\n",
    "    line = line.rstrip()\n",
    "    # check id the line contains a > chracter, this is the difference between header and sequence\n",
    "    if line.count(\">\") > 0: # if there is more than 0 > itis a header as sequence lines do not have this character\n",
    "        head_line = line.split(\"|\") # split by straight dash\n",
    "        bsm_id.append(head_line[0]) # add the id (entry 1, index 0) to the id list\n",
    "        bsm_start.append(head_line[1]) # add the start (entry 2, index 1) to the start list \n",
    "        bsm_stop.append(head_line[2]) # add the end (entry 3, index 2) to the end list\n",
    "        bsm_chr.append(head_line[3]) # add the chromosome (entry 4, index 3) to the chr list\n",
    "    else:\n",
    "        bsm_seq.append(line) # if the line is not a header, the whole line can be added to the sequence list\n",
    "\n",
    "bsm_file.close()\n",
    "        \n",
    "print(bsm_id[0:10])\n",
    "print(bsm_start[0:10])\n",
    "print(bsm_stop[0:10])\n",
    "print(bsm_chr[0:10])\n",
    "print(bsm_seq[0:10])\n",
    "\n",
    "[len(bsm_id),\n",
    "len(bsm_start),\n",
    "len(bsm_stop),\n",
    "len(bsm_chr),\n",
    "len(bsm_seq)]\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 2.4 Concatenate (glue) the sequence lines together\n",
    "# when we check the length we see that we have equal numbers of id, start, stop and chromosome entries.\n",
    "# we have much more sequenc entries because of the multiple lines per sequence, when we glue them together this should be fine\n",
    "\n",
    "\n",
    "bsm_file = open(\"Bolivian_squirrel_monkey_genes_mart_export.txt\",\"r\")\n",
    "\n",
    "# we need to make a separate list for id, start, stop and chromosome and therefor also for the sequence\n",
    "bsm_id = []\n",
    "bsm_start = []\n",
    "bsm_stop = []\n",
    "bsm_chr = []\n",
    "bsm_seq = []\n",
    "\n",
    "for line in bsm_file:\n",
    "    # first remove the \\n\n",
    "    line = line.rstrip()\n",
    "    # check id the line contains a > chracter, this is the difference between header and sequence\n",
    "    if line.count(\">\") > 0: # if there is more than 0 > it is a header as sequence lines do not have this character\n",
    "        head_line = line.split(\"|\") # split by straight dash\n",
    "        bsm_id.append(head_line[0]) # add the id (entry 1, index 0) to the id list\n",
    "        bsm_start.append(head_line[1]) # add the start (entry 2, index 1) to the start list \n",
    "        bsm_stop.append(head_line[2]) # add the end (entry 3, index 2) to the end list\n",
    "        bsm_chr.append(head_line[3]) # add the chromosome (entry 4, index 3) to the chr list\n",
    "    else:\n",
    "##############################################################################################################################\n",
    "## here we need to make sure we end up with one sequence per gene\n",
    "        ## to know if a gene has a sequence entry already we can check the length of the id list and the chromosome list\n",
    "        if len(bsm_seq) < len(bsm_id): \n",
    "            bsm_seq += [\"\"] ## if the seq lenght is shorter we can add an empty item to the list\n",
    "        \n",
    "        bsm_seq[len(bsm_id)-1] += line # the we can concatenate the rest of the sequence to this item \n",
    "##############################################################################################################################\n",
    "bsm_file.close()\n",
    "        \n",
    "print(bsm_id[0:10])\n",
    "print(bsm_start[0:10])\n",
    "print(bsm_stop[0:10])\n",
    "print(bsm_chr[0:10])\n",
    "print(bsm_seq[0:10])\n",
    "\n",
    "[len(bsm_id),\n",
    "len(bsm_start),\n",
    "len(bsm_stop),\n",
    "len(bsm_chr),\n",
    "len(bsm_seq)]\n",
    "        \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 3 Getting data from a file in a dictionary\n",
    "\n",
    "Having the data in separate lists is great, but not ideal. The data has to be in the same order and there is no link between the entries. Also we hav multiple lists, whereas one container might be easier.\n",
    "\n",
    "A Dictonary links the data entries and makes processing easier and mor reproducible.\n",
    "\n",
    "3.1) Start by making an empty dictionary and with getting the data from the file line by line. The line of data is not stored directly in the dictionary, but processed first. Strip the line of the end of line character (\\n) and store it in a new variable."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 3.1)\n",
    "\n",
    "# Make empty dictionary\n",
    "bsm_genes_dict = {}\n",
    "\n",
    "# open file connection\n",
    "bsm_file = open(\"Bolivian_squirrel_monkey_genes_mart_export.txt\",\"r\")\n",
    "\n",
    "## loop over file line-by-line\n",
    "for line in bsm_file:\n",
    "    ## strip end-of-line and store in new variable\n",
    "    tmp_line = line.rstrip() ## .rstrip() can be used to remove the \\n character from the end of the line\n",
    "## close file connection \n",
    "bsm_file.close()\n",
    "\n",
    "print(bsm_genes_dict)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "3.2) Copy the script from 3.1)\n",
    "\n",
    "We need to process the data line per line befor we can enter it in a dictionary. Let's start with the header. A header can be recognised by the > character a the start of a line. \n",
    "Look up how we checked for a line containing a header using an if statement and counting ofr a certain character\n",
    "\n",
    "The data stored in the header is the gene id, start, stop and chromosome.\n",
    "\n",
    "The gene id will make an good dictionary key, the other data can be values (in a list) or in their own dictionary.\n",
    "\n",
    "For example\n",
    "{id:{\"start\":0,\"stop\":1,\"chr\":\"1\",\"seq\":\"\"}\n",
    "\n",
    "Add an empty \"seq\" key as we need to get that info next itteration\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 3.2)\n",
    "\n",
    "# Make empty dictionary\n",
    "bsm_genes_dict = {}\n",
    "\n",
    "# open file connection\n",
    "bsm_file = open(\"Bolivian_squirrel_monkey_genes_mart_export.txt\",\"r\")\n",
    "\n",
    "## loop over file line-by-line\n",
    "for line in bsm_file:\n",
    "    ## strip end-of-line and store in new variable\n",
    "    tmp_line = line.rstrip() ## .rstrip() can be used to remove the \\n character from the end of the line\n",
    "    if tmp_line.count(\">\") > 0: # if there is more than 0 > it is a header as sequence lines do not have this character\n",
    "        ## split the header line by straight dash\n",
    "        head_line = line.split(\"|\") # split by straight dash\n",
    "        # get the data in a dictionary with the gene id as key\n",
    "        bsm_genes_dict[head_line[0]] = {\"start\":head_line[1],\"stop\":head_line[2],\"chr\":head_line[3],\"seq\":\"\"}\n",
    "## close file connection \n",
    "bsm_file.close()\n",
    "\n",
    "print(bsm_genes_dict)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "3.3) Now that we have our header information in a dictionary it is time to add the sequence info\n",
    "\n",
    "Copy the script of 3.2, add an else statement  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 3.3)\n",
    "\n",
    "# Make empty dictionary\n",
    "bsm_genes_dict = {}\n",
    "\n",
    "# open file connection\n",
    "bsm_file = open(\"Bolivian_squirrel_monkey_genes_mart_export.txt\",\"r\")\n",
    "\n",
    "## loop over file line-by-line\n",
    "for line in bsm_file:\n",
    "    ## strip end-of-line and store in new variable\n",
    "    tmp_line = line.rstrip() ## .rstrip() can be used to remove the \\n character from the end of the line\n",
    "    if tmp_line.count(\">\") > 0: # if there is more than 0 > it is a header as sequence lines do not have this character\n",
    "        ## split the header line by straight dash\n",
    "        head_line = tmp_line.split(\"|\") # split by straight dash\n",
    "        # get the data in a dictionary with the gene id as key\n",
    "        bsm_genes_dict[head_line[0]] = {\"start\":head_line[1],\"stop\":head_line[2],\"chr\":head_line[3],\"seq\":\"\"}\n",
    "    ## add an else stament\n",
    "    else:\n",
    "        ## add sequence to \"seq\" key at the right gene id make sure you concatenate the sequences when they are one multiple lines\n",
    "        bsm_genes_dict[head_line[0]][\"seq\"] += tmp_line\n",
    "## close file connection \n",
    "bsm_file.close()\n",
    "\n",
    "print(bsm_genes_dict[list(bsm_genes_dict.keys())[7]])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 4 Exploring the data\n",
    "\n",
    "Now that the data from the file is stored in lists and/or dictionary we can try to answer some questions\n",
    "\n",
    "4.1) Howmany genes are on each chromosome? Plot the result in a histogram\n",
    "\n",
    "4.2) Howmany 'Sequence unavilable' are in the dataset\n",
    "\n",
    "4.3) Plot an histogram with the gene lengths. What is the median gene length for each chromosome?     \n",
    "\n",
    "4.4) Plot a histogram with the gene coding lengths. What is the median coding length for each chromosome?\n",
    "\n",
    "4.5) What is the codon usage for each chromosome?\n",
    "\n",
    "4.6) Plot a histogram of a the usage of a codon as percentage of total basepairs for each gene.\n",
    "\n",
    "4.7) Plot for one chromosome the chromosome start of gene vs the gene length. And a histogram of the start positions\n",
    "\n",
    "4.8) Translate the the coding sequences into proteins and store them in the dictionary\n",
    "\n",
    "4.9) Make a function that reports the peptides and their length after digestion of the protein on a given amino-acid sequence\n",
    "\n",
    "4.10) Make a function that takes the dictionary with genome data and\n",
    "        # outputs a dictinonary with gene sequences per chromosome, exclude \"Sequence unavailable\"\n",
    "        # set a threshold for the minimum genes per chromosome to be taken up in the dictionary, \n",
    "                # for example only chromosomes with more than 300 genes\n",
    "         \n",
    "4.11) make a function that shows the distributio of GC contents of multiple groups of genes. \n",
    "        For example the genes per chromosome from 4.10)\n",
    "\n",
    "4.12) Make a function that compares the GC content of two groups of genes in a figure\n",
    "\n",
    "4.13) Can you make up your own question and visualisation for the results?\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 4.1) Howmany genes are on each chromosome (X, Y, MT, ...)?\n",
    "chr_gene_count = {}\n",
    "for ids in bsm_genes_dict:\n",
    "    new_id = bsm_genes_dict[ids][\"chr\"].replace(\".\",\"_\")\n",
    "    # print(new_id)\n",
    "    if new_id in chr_gene_count:\n",
    "        chr_gene_count[new_id] += 1\n",
    "    else:\n",
    "        chr_gene_count[new_id] = 1\n",
    "        \n",
    "chr_gene_count    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "\n",
    "plt.hist(chr_gene_count.values(),bins=100,color=\"red\")\n",
    "plt.title(\"Gene counts per chromosome\")\n",
    "plt.ylabel(\"Frequency\")\n",
    "plt.xlabel(\"Genes\")\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 4.2) Howmany 'Sequence unavailable' are in the dataset\n",
    "\n",
    "su_count = 0\n",
    "for ids in bsm_genes_dict:\n",
    "    if bsm_genes_dict[ids][\"seq\"] == \"Sequence unavailable\":\n",
    "        su_count +=1\n",
    "        \n",
    "print(su_count)\n",
    "\n",
    "len(bsm_genes_dict.keys())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 4.3) Plot an histogram with the gene lengths. \n",
    "# Set gene lengths > 100000 bp to 100000\n",
    "# What is the median gene length for each chromosome?    \n",
    "gene_lengths = []\n",
    "gl_per_chr = {}\n",
    "for ids in bsm_genes_dict:\n",
    "    # print(int(bsm_genes_dict[ids][\"stop\"]))\n",
    "    gl = abs(int(bsm_genes_dict[ids][\"stop\"]) - int(bsm_genes_dict[ids][\"start\"])) + 1 \n",
    "    if gl > 100000:\n",
    "        gl = 100000\n",
    "    gene_lengths += [gl]\n",
    "    bsm_genes_dict[ids][\"length\"] = gl\n",
    "    new_id = bsm_genes_dict[ids][\"chr\"].replace(\".\",\"_\")\n",
    "    if new_id in gl_per_chr:\n",
    "        gl_per_chr[new_id] += [gl]\n",
    "    else:\n",
    "        gl_per_chr[new_id] = [gl]            \n",
    "\n",
    "plt.hist(gene_lengths,bins=100,color=\"magenta\")  \n",
    "plt.show()\n",
    "# print(gene_lengths)\n",
    "# gl_per_chr\n",
    "\n",
    "import statistics as stats\n",
    "\n",
    "med_gl_per_chr = {}\n",
    "for chr_ids in gl_per_chr:\n",
    "    med_gl_per_chr[chr_ids] = stats.median(gl_per_chr[chr_ids])\n",
    "# print(med_gl_per_chr)\n",
    "\n",
    "plt.bar(med_gl_per_chr.keys(),med_gl_per_chr.values())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 4.4) Plot a histogram with the gene coding lengths. \n",
    "# What is the median coding length for each chromosome?\n",
    "coding_lengths = []\n",
    "cl_per_chr = {}\n",
    "for ids in bsm_genes_dict:\n",
    "    # print(int(bsm_genes_dict[ids][\"stop\"]))\n",
    "    cl = len(bsm_genes_dict[ids][\"seq\"])\n",
    "    if cl > 100000:\n",
    "        cl = 100000\n",
    "    coding_lengths += [cl]\n",
    "    bsm_genes_dict[ids][\"coding\"] = cl\n",
    "    new_id = bsm_genes_dict[ids][\"chr\"].replace(\".\",\"_\")\n",
    "    if new_id in cl_per_chr:\n",
    "        cl_per_chr[new_id] += [cl]\n",
    "    else:\n",
    "        cl_per_chr[new_id] = [cl]            \n",
    "\n",
    "plt.hist(coding_lengths,bins=100,color=\"magenta\")  \n",
    "plt.show()\n",
    "# print(gene_lengths)\n",
    "# gl_per_chr\n",
    "\n",
    "import statistics as stats\n",
    "\n",
    "med_cl_per_chr = {}\n",
    "for chr_ids in cl_per_chr:\n",
    "    med_cl_per_chr[chr_ids] = stats.median(cl_per_chr[chr_ids])\n",
    "# print(med_gl_per_chr)\n",
    "\n",
    "plt.bar(med_cl_per_chr.keys(),med_cl_per_chr.values())\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 4.5) What is the codon usage for each chromosome?\n",
    "import dna_tools as tools \n",
    "\n",
    "tools.translate_dna(\"AATTGGCTG\")\n",
    "\n",
    "cod_per_chr = {}\n",
    "for ids in bsm_genes_dict:\n",
    "    # print(int(bsm_genes_dict[ids][\"stop\"]))\n",
    "    if bsm_genes_dict[ids][\"seq\"] != \"Sequence unavailable\":\n",
    "        coduse = tools.get_counts(bsm_genes_dict[ids][\"seq\"],3)\n",
    "        new_id = bsm_genes_dict[ids][\"chr\"].replace(\".\",\"_\")\n",
    "        if new_id in cod_per_chr:\n",
    "            for codid in coduse:\n",
    "                if codid in cod_per_chr[new_id]:\n",
    "                    cod_per_chr[new_id][codid] += coduse[codid]\n",
    "                else:\n",
    "                    cod_per_chr[new_id][codid] = coduse[codid]\n",
    "        else:\n",
    "            cod_per_chr[new_id] = coduse    \n",
    "            \n",
    "cod_per_chr            \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 4.6) Plot a histogram of a the usage of a codon as percentage of total basepairs for each gene.\n",
    "import dna_tools as tools \n",
    "\n",
    "cod_per_chr = {}\n",
    "for ids in bsm_genes_dict:\n",
    "    # print(int(bsm_genes_dict[ids][\"stop\"]))\n",
    "    if bsm_genes_dict[ids][\"seq\"] != \"Sequence unavailable\":\n",
    "        use_seq = bsm_genes_dict[ids][\"seq\"]\n",
    "        coduse = tools.get_counts(use_seq,3)\n",
    "        new_id = bsm_genes_dict[ids][\"chr\"].replace(\".\",\"_\")\n",
    "        if new_id in cod_per_chr:\n",
    "            for codid in coduse:\n",
    "                if codid in cod_per_chr[new_id]:\n",
    "                    cod_per_chr[new_id][codid] += coduse[codid]/len(use_seq)\n",
    "                else:\n",
    "                    cod_per_chr[new_id][codid] = coduse[codid]/len(use_seq)\n",
    "        else:\n",
    "            for codid in coduse:\n",
    "                coduse[codid] = coduse[codid]/len(use_seq)    \n",
    "                cod_per_chr[new_id] = coduse    \n",
    "            \n",
    "cod_per_chr            \n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 4.7) Plot for one chromosome the chromosome start of gene vs the gene length. And a histogram of the start positions\n",
    "import matplotlib.pyplot as plt\n",
    "gene_lengths = []\n",
    "gl_per_chr = {}\n",
    "start_per_chr = {}\n",
    "for ids in bsm_genes_dict:\n",
    "    # print(int(bsm_genes_dict[ids][\"stop\"]))\n",
    "    gl = abs(int(bsm_genes_dict[ids][\"stop\"]) - int(bsm_genes_dict[ids][\"start\"])) + 1 \n",
    "    if gl > 100000:\n",
    "        gl = 100000\n",
    "    gene_lengths += [gl]\n",
    "    bsm_genes_dict[ids][\"length\"] = gl\n",
    "    new_id = bsm_genes_dict[ids][\"chr\"].replace(\".\",\"_\")\n",
    "    if new_id in gl_per_chr:\n",
    "        gl_per_chr[new_id] += [gl]\n",
    "        start_per_chr[new_id] += [int(bsm_genes_dict[ids][\"start\"])]\n",
    "    else:\n",
    "        gl_per_chr[new_id] = [gl]\n",
    "        start_per_chr[new_id] = [int(bsm_genes_dict[ids][\"start\"])]\n",
    "\n",
    "# print(gene_lengths)\n",
    "# print(start_per_chr)\n",
    "x = start_per_chr[list(gl_per_chr.keys())[0]]\n",
    "y = gl_per_chr[list(gl_per_chr.keys())[0]]\n",
    "plt.scatter(x,y)\n",
    "plt.show()\n",
    "plt.hist(x,bins=100)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 4.8) Translate the the coding sequences into proteins\n",
    "import dna_tools as tools \n",
    "\n",
    "tools.translate_dna(\"AATTGGCTG\") # check if the translate function works \n",
    "\n",
    "for ids in bsm_genes_dict:\n",
    "    bsm_genes_dict[ids][\"prot\"] = tools.translate_dna(bsm_genes_dict[ids][\"seq\"])\n",
    "    \n",
    "print(bsm_genes_dict[list(bsm_genes_dict.keys())[7]]) \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 4.9) Make a function that reports the peptides and their length after digestion of the protein on a given amino-acid sequence\n",
    "\n",
    "    # hint: use x to indicate the exact cutting site \n",
    "def dig_prot(prot_seq,aa_dig_pattern):\n",
    "    if \"re\" not in dir():\n",
    "        import re as re\n",
    "    aa_dig_seq = aa_dig_pattern.replace(\"x\",\"\")    \n",
    "    tmp_prot = prot_seq.replace(aa_dig_seq,aa_dig_pattern)\n",
    "    fragments = tmp_prot.split(\"x\")\n",
    "    frag_len = []\n",
    "    for frag in fragments:\n",
    "        frag_len += [len(frag)]\n",
    "    return(fragments,frag_len)\n",
    "    \n",
    "## test the function    \n",
    "use_protein = bsm_genes_dict[list(bsm_genes_dict.keys())[7]][\"prot\"]\n",
    "use_dig = \"TxSA\"\n",
    "dig_prot(use_protein,use_dig)    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 4.10 Make a function that takes the dictionary with genome data and\n",
    "        # outputs a dictinonary with gene sequences per chromosome, exclude \"Sequence unavailable\"\n",
    "        # set a threshold for the minimum genes per chromosome to be taken up in the dictionary, \n",
    "                # for example only chromosomes with more than 300 genes\n",
    "## example {\"chr1\":[\"atacgat\",\"gtgatatacatct\",\"ccggaattt\",...],\"chr2\":[\"atg\",\"ccc\",\"ggt\",...],...}\n",
    "\n",
    "def seq_by_chromosome(input_dict,gene_thr = 300):\n",
    "    ## get the sequences in a dcitionary, with the chromosome names as keys\n",
    "    chr_seq_dict = {}\n",
    "    for ids in input_dict:\n",
    "        chr_key = input_dict[ids][\"chr\"]\n",
    "        tmp_seq = input_dict[ids][\"seq\"]\n",
    "        if tmp_seq not in \"Sequence unavailable\": # check if the sequence in not Sequence unavailable\n",
    "            if chr_key in chr_seq_dict:\n",
    "                chr_seq_dict[chr_key] += [tmp_seq]\n",
    "            else:\n",
    "                chr_seq_dict[chr_key] = [tmp_seq]\n",
    "                \n",
    "    chr_keys = list(chr_seq_dict.keys())\n",
    "    print(chr_keys)\n",
    "    new_keys = []\n",
    "    for chrid in chr_keys:\n",
    "        if len(chr_seq_dict[chrid]) > gene_thr:\n",
    "            new_keys += [chrid]\n",
    "            print(new_keys)\n",
    "    \n",
    "    chr_seq_dict2 = {}         \n",
    "    for chrid in new_keys:\n",
    "        chr_seq_dict2[chrid] = chr_seq_dict[chrid]\n",
    "    \n",
    "    return(chr_seq_dict2)\n",
    "## test function\n",
    "ok = seq_by_chromosome(bsm_genes_dict,200)\n",
    "print(ok[list(ok.keys())[0]])                     \n",
    "                     "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 4.11 make a function that shows the distributio of GC contents of multiple groups of genes. \n",
    "    # For example the genes per chromosome\n",
    "    # hint you can use plt.violinplot\n",
    "    \n",
    "def violin_GC(input_dict):\n",
    "    ## check if the matplotlib.pyplot is loaded\n",
    "    if \"plt\" not in dir():\n",
    "        import matplotlib.pyplot as plt\n",
    "    \n",
    "    if  \"ax\" not in dir():\n",
    "        import matplotlib.axes as ax\n",
    "    \n",
    "    ## make function to count GC content for each gene\n",
    "    def get_cg_content (gene_seq):\n",
    "        at = gene_seq.count(\"A\") + gene_seq.count(\"T\")\n",
    "        cg = gene_seq.count(\"C\") + gene_seq.count(\"G\")\n",
    "        if ((at+cg) > 0):\n",
    "            gc_content = cg/(at+cg)\n",
    "        else:\n",
    "            gc_content = 0\n",
    "        return gc_content\n",
    "    ## get chromosome labels so they can be used later and stay in the same order\n",
    "    chr_labels = list(input_dict.keys())\n",
    "    ## make list to collect the lists of gc contents\n",
    "    gc_list = []\n",
    "    for chrid in chr_labels:\n",
    "        tmp_gc = []\n",
    "        for seq in input_dict[chrid]: \n",
    "            tmp_gc += [get_cg_content(seq)]\n",
    "        gc_list += [tmp_gc]    \n",
    "    \n",
    "    plt.violinplot(gc_list,widths=0.9) \n",
    "    plt.boxplot(gc_list,widths=0.2)\n",
    "    plt.xticks(range(1,len(chr_labels)+1), chr_labels,rotation =45)\n",
    "    plt.ylabel(\"%GC\")\n",
    "    plt.xlabel(\"Chromosome\")\n",
    "    plt.show()\n",
    "    \n",
    "violin_GC(ok)   \n",
    "    \n",
    "len(ok.keys())   \n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 4.12 Make a function that compares the GC content of two groups of genes in a figure\n",
    "\n",
    "def compare_GC(group1,group2):\n",
    "    if \"plt\" not in dir():\n",
    "        import matplotlib.pyplot as plt\n",
    "    gc_content1 = []\n",
    "    gc_content2 = []\n",
    "    for gene in group1:\n",
    "        at = gene.count(\"A\") + gene.count(\"T\")\n",
    "        gc = gene.count(\"C\") + gene.count(\"G\")\n",
    "        if ((at+gc) > 0):\n",
    "            gc_content1 += [gc/(at+gc)]\n",
    "        else:\n",
    "            gc_content2 += [0]\n",
    "    for gene in group1:\n",
    "        at = gene.count(\"A\") + gene.count(\"T\")\n",
    "        gc = gene.count(\"C\") + gene.count(\"G\")\n",
    "        if ((at+gc) > 0):\n",
    "            gc_content2 += [gc/(at+gc)]\n",
    "        else:\n",
    "            gc_content2 += [0]\n",
    "    \n",
    "    plt.hist([gc_content1,gc_content2],bins=100) \n",
    "    plt.show()\n",
    "    plt.violinplot([gc_content1,gc_content2]) \n",
    "    \n",
    "## test function    \n",
    "\n",
    "## make two groups of genes baseed on length\n",
    "grp1 = []\n",
    "grp2 = []\n",
    "for ids in bsm_genes_dict:\n",
    "    gseq = bsm_genes_dict[ids][\"seq\"]\n",
    "    if len(gseq) > 60 & len(gseq) < 601:\n",
    "        grp1 += [gseq]\n",
    "    elif len(gseq) > 600:\n",
    "        grp2 += [gseq]\n",
    "        \n",
    "compare_GC(grp1,grp2)            "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 4.13) Can you make up your own question and visualisation for the results?\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 5 Can you make a function to\n",
    "\n",
    "5.1) load the data from a genome file with the same structure into a dictionary\n",
    "\n",
    "5.2) test the function using the Dolphin_genes_mart_export.txt file\n",
    "\n",
    "5.3) A function that extracts genome features such as, no. of genes, average gene length, coding length,\n",
    " make figures of the gene length distribution, coding length distrbution, gc content distribution,\n",
    " coding length vs gene length, gc vs coding length\n",
    "\n",
    "5.4) test the function using the Dolphin_genes_mart_export.txt file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 5.1) load the data from a genome file with the same structure into a dictionary\n",
    "def get_genome_data(use_file):\n",
    "    \n",
    "    genome_file = open(use_file,\"r\")\n",
    "\n",
    "    genome_genes = {}\n",
    "    for line in genome_file:\n",
    "        line = line.rstrip()\n",
    "        if ( len(line) > 0):\n",
    "            data = line.split(\"|\")\n",
    "            if (data[0][0] == \">\"):\n",
    "                gene_info = {\"start\":int(data[1]),\"stop\":int(data[2]),\"contig\":data[3],\"sequence\":\"\"}\n",
    "                use_key = data[0]\n",
    "                genome_genes[use_key] = gene_info\n",
    "            else:\n",
    "                genome_genes[use_key][\"sequence\"] += data[0]\n",
    "\n",
    "    genome_file.close()\n",
    "    return genome_genes\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 5.2) test the function using the Dolphin_genes_mart_export.txt file\n",
    "dol_genes = get_genome_data(use_file = \"Dolphin_genes_mart_export.txt\")\n",
    "print(dol_genes[list(dol_genes.keys())[2]])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 5.3) A function that extracts genome features such as, no. of genes, average gene length, coding length,\n",
    "# make figures of the gene length distribution, coding length distrbution, gc content distribution,\n",
    "# coding length vs gene length, gc vs coding length\n",
    "\n",
    "def get_cg_content (gene_seq):\n",
    "    at = gene_seq.count(\"A\") + gene_seq.count(\"T\")\n",
    "    cg = gene_seq.count(\"C\") + gene_seq.count(\"G\")\n",
    "    if ((at+cg) > 0):\n",
    "        gc_content = cg/(at+cg)\n",
    "    else:\n",
    "        gc_content = 0\n",
    "    return gc_content\n",
    "\n",
    "\n",
    "def get_start_stop_codon (use_genome):\n",
    "    import matplotlib.pyplot as plt\n",
    "    start_codon = {}\n",
    "    stop_codon = {}\n",
    "    for seq_key in use_genome:\n",
    "        tmp_seq = use_genome[seq_key][\"sequence\"]\n",
    "        tmp_start = tmp_seq[0:3]\n",
    "        if(tmp_start not in start_codon.keys()):\n",
    "            start_codon[tmp_start] = 1\n",
    "        else:\n",
    "            start_codon[tmp_start] += 1\n",
    "\n",
    "        tmp_stop = tmp_seq[(len(tmp_seq)-3):(len(tmp_seq)+1)]\n",
    "        if(tmp_stop not in stop_codon.keys()):\n",
    "            stop_codon[tmp_stop] = 1\n",
    "        else:\n",
    "            stop_codon[tmp_stop] += 1\n",
    "\n",
    "    plt.subplots(ncols=2,dpi=150)\n",
    "    plt.bar(start_codon.keys(),start_codon.values())\n",
    "    plt.bar(stop_codon.keys(),stop_codon.values())\n",
    "    plt.xticks(rotation=\"vertical\",fontsize=6) ;\n",
    "    plt.show()\n",
    "    \n",
    "\n",
    "def show_genome_features(use_genome,\n",
    "                        show_no_of_genes = True,\n",
    "                        show_average_gene_length = True,\n",
    "                        show_average_coding_length = True,\n",
    "                        plot_gene_length_distri = True,\n",
    "                        plot_coding_length_distri = True,\n",
    "                        plot_coding_vs_gene_length = True,\n",
    "                        plot_gc_content_distri = True,\n",
    "                        plot_gc_vs_coding_len = True):\n",
    "    ### import modules\n",
    "    import matplotlib.pyplot as plt\n",
    "    \n",
    "    ## extract features\n",
    "        ### gene lengths\n",
    "    gene_len = []\n",
    "    for seq_key in use_genome:\n",
    "        tmp = [use_genome[seq_key][\"stop\"]-use_genome[seq_key][\"start\"]]\n",
    "        gene_len += tmp\n",
    "        ### coding lengths\n",
    "    coding_len = []\n",
    "    for seq_key in use_genome:\n",
    "        coding_len += [len(use_genome[seq_key][\"sequence\"])]    \n",
    "        ### gc content\n",
    "    gc_per_gene = []    \n",
    "    for seq_key in use_genome:\n",
    "        gc_per_gene += [get_cg_content(use_genome[seq_key][\"sequence\"])]\n",
    "    \n",
    "    \n",
    "    ### show no of genes\n",
    "    if(show_no_of_genes):\n",
    "        print(\"This genome consists of \" + str(len(use_genome))  + \" genes\")\n",
    "    \n",
    "    ### show average gene length\n",
    "    if(show_average_gene_length):\n",
    "        print(\"The average gene length is \" + str(sum(gene_len)/len(gene_len)) + \" basepairs\")   \n",
    "        \n",
    "    ### show average coding length    \n",
    "    if(show_average_coding_length):\n",
    "        print(\"The average coding region is \" + str(sum(coding_len)/len(coding_len)) + \" basepairs\")   \n",
    "    \n",
    "    ### plot gene length distribution\n",
    "    if(plot_gene_length_distri):\n",
    "        plt.hist(gene_len,bins=1000)\n",
    "        plt.title(\"Gene Length\")\n",
    "        plt.xlim(0,20000)\n",
    "        plt.show()\n",
    "     \n",
    "    ### plot coding length distribution\n",
    "    if(plot_coding_length_distri):\n",
    "        plt.hist(coding_len,bins=1000)\n",
    "        plt.title(\"Coding Length\")\n",
    "        plt.xlim(0,20000)        \n",
    "        plt.show()\n",
    "        \n",
    "    ### plot coding vs gene length\n",
    "    if(plot_coding_vs_gene_length):\n",
    "        plt.plot(gene_len,coding_len,linestyle=\"\",marker=\".\")\n",
    "        plt.xlim(0,20000)\n",
    "        plt.ylim(0,20000)\n",
    "        plt.xlabel(\"Gene Length\")\n",
    "        plt.ylabel(\"Coding Length\")\n",
    "        plt.show()\n",
    "    \n",
    "    ### plot gc_content_distri\n",
    "    if(plot_gc_content_distri):\n",
    "        plt.hist(gc_per_gene,bins=200)\n",
    "        plt.xlim(0.1,0.9)\n",
    "        plt.ylim(0,1000)\n",
    "        plt.title(\"GC content\")\n",
    "        plt.show()\n",
    "     \n",
    "    ### plot gc vs coding length\n",
    "    if(plot_gc_vs_coding_len):\n",
    "        plt.plot(coding_len,gc_per_gene,linestyle=\"\",marker=\".\")\n",
    "        plt.xlabel(\"Gene Length\")\n",
    "        plt.ylabel(\"GC\")\n",
    "        plt.xlim(0,20000)\n",
    "        plt.show()\n",
    "        \n",
    "    get_start_stop_codon(use_genome)\n",
    "      \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 5.4 Test the function with  Dolphin_genes_mart_export.txt file\n",
    "dol_genes = get_genome_data(use_file = \"Dolphin_genes_mart_export.txt\")\n",
    "show_genome_features(dol_genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
